README.md
# cordova-plugin-googleplace

A Cordova plugin to use the Google Places SDK.

**NOTE: this is a preliminary version, as of now only iOS is supported and Android is in the making**

## Getting Started

This plugin allows to user the Google Places SDK from Cordova, for building autocomplete UIs for locations.

### Installing

```
$ SHELL COMMAND TO INSTALL
$ cordova plugin add cordova-plugin-googleplaces --variable API_KEY_FOR_IOS="XXXX"
```

### Usage

For more details you can generate the Javascript docs with `yarn doc`:

```sh
$ yarn doc
yarn run v1.1.0
$ docco www/GooglePlaces.js
docco: www/GooglePlaces.js -> docs/GooglePlaces.html
```

#### `currentPlace`

You can call the `currentPlace` method to find information on the current locatoin.

This method **requires** that the user has enabled geolocation in the app. 
To to so, use cordova-plugin-geolocation (or equivalent) before calling this method.

```javascript
// Authorize access to the current position using cordova-plugin-geolocation
navigator.geolocation.getCurrentPosition(function(pos) {

  // retreive the current place
  cordova.plugins.GooglePlaces.currentPlace(
    // place contains the API result
    place => {
      console.log(place);
      //   {
      //    place: {
      //      name: "some place name",
      //      placeID: "XXXXX"
      //    },
      //    likehood: 0.87 // <= means 87% accurate
      //   }
    },
    err => console.log(err)
  );
});

```

#### `autocompleteQuery`

The `autocompleteQuery(query, [bounds], [filter], success, failure)` method find candidates ("predictions") given an input
query string:

```javascript
cordova.plugins.GooglePlaces.autocompleteQuery("22 some random address",
  // results contains an array of predictions, the first being the most pobable
  results => {
    console.log(res);
    //   {
    //     fullText: "description of the place",
    //     primaryText: "partial description of the place",
    //     secondaryText: "partial description of the place",
    //     placeID: "XXXXX",
    //     types: [ "a", "list", "of", "types", "for", "the", "result" ]
    //   }
  },
  err => console.log(err),
);

```

This method can take optional arguments:

- `bounds` defines a regions to limit the search to.
  
  It should be defined as a "coordinate region" object such as:
  
  ```javascript
  {
   northEast: {
     latitude: 1.234,
     longitude: 5,667
   },
   southWest: {
     latitude: 1.234,
     longitude: 5,667
   }
  }
  ```

- `filter` defines a filter to limit the results to a specific region.

  Such a filter is given by a filter type taken from the `GooglePlaces.AutocompleteFilterTypes` and an (optional) country:

  ```javascript
  {
    filter: google.plugins.GooglePlaces.AutocompleteFilterTypes.NoFilter,
    country: "FR" // <= this is optional
  }
  ```

  Several values are available for the filter type:
  - `AutocompleteFilterTypes.NoFilter` is an empty filter; all results are returned.
  - `AutocompleteFilterTypes.Geocode` returns only autocomplete results with a precise address. Use this type when you know the user is looking for a fully specified address.
  - `AutocompleteFilterTypes.Address` returns only places that are businesses.
  - `AutocompleteFilterTypes.Establishment` returns only places that are businesses.
  - `AutocompleteFilterTypes.Region` returns only places that match one of the following types: `locality`, `sublocality`, `postal_code`, `country`, `administrative_area_level_1`, `administrative_area_level_2`
  - `AutocompleteFilterTypes.City` returns only results matching `locality` or `administrative_area_level_3`.

#### `pickPlace`

`pickplace([bounds], success, [failure])` displays the native UI for picking a nearby place.

This method **requires** that the user has enabled geolocation in the app. 
To to so, use cordova-plugin-geolocation (or equivalent) before calling this method.

```javascript
cordova.plugins.GooglePlaces.pickPlace(
  // place contains complete place information on the selected target
  place => {
    console.log(res);
    //   {
    //      name: "some place name",
    //      placeID: "XXXXX",
    //      // and lots of other fields, depending on the place info available
    //   }
  },
  err => console.log(err),
);
```

The optional `bounds` argument is defined as explained above:
  
```javascript
{
 northEast: {
   latitude: 1.234,
   longitude: 5,667
 },
 southWest: {
   latitude: 1.234,
   longitude: 5,667
 }
}
```

#### `showPlaceAutocomplete`

Displays the native UI for place autocompletion.

```javascript
cordova.plugins.GooglePlaces.showPlaceAutocomplete(
  // place contains complete place information on the selected target
  place => {
    console.log(res);
    //   {
    //      name: "some place name",
    //      placeID: "XXXXX",
    //      // and lots of other fields, depending on the place info available
    //   }
  },
  err => console.log(err),
);

```

## Contributing

Feel free to contribute anytime !

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

## TODO

* [ ] Add Android support



CHANGELOG.md
# Changelog

## [Unreleased]

- Add basic file structure for the plugin
- Add JS tools:
  - rollup w/ babel for transpiling and packaging
  - esdoc for generating the user's documentation
- Add [`www/GooglePlaces.js`](www/GooglePlaces.js) with the source 
  code for calling the non-yet-existing native plugin



www/GooglePlaces.dist.js
"use strict";function getRawTag(e){var o=hasOwnProperty.call(e,symToStringTag$1),t=e[symToStringTag$1];try{e[symToStringTag$1]=void 0;var n=!0}catch(e){}var r=nativeObjectToString.call(e);return n&&(o?e[symToStringTag$1]=t:delete e[symToStringTag$1]),r}function objectToString(e){return nativeObjectToString$1.call(e)}function baseGetTag(e){return null==e?void 0===e?undefinedTag:nullTag:symToStringTag&&symToStringTag in Object(e)?_getRawTag(e):_objectToString(e)}function isObject(e){var o=typeof e;return null!=e&&("object"==o||"function"==o)}function isFunction(e){if(!isObject_1(e))return!1;var o=_baseGetTag(e);return o==funcTag||o==genTag||o==asyncTag||o==proxyTag}function _classCallCheck(e,o){if(!(e instanceof o))throw new TypeError("Cannot call a class as a function")}var commonjsGlobal="undefined"!=typeof window?window:"undefined"!=typeof global?global:"undefined"!=typeof self?self:{},freeGlobal="object"==typeof commonjsGlobal&&commonjsGlobal&&commonjsGlobal.Object===Object&&commonjsGlobal,_freeGlobal=freeGlobal,freeSelf="object"==typeof self&&self&&self.Object===Object&&self,root=_freeGlobal||freeSelf||Function("return this")(),_root=root,Symbol$1=_root.Symbol,_Symbol=Symbol$1,objectProto=Object.prototype,hasOwnProperty=objectProto.hasOwnProperty,nativeObjectToString=objectProto.toString,symToStringTag$1=_Symbol?_Symbol.toStringTag:void 0,_getRawTag=getRawTag,objectProto$1=Object.prototype,nativeObjectToString$1=objectProto$1.toString,_objectToString=objectToString,nullTag="[object Null]",undefinedTag="[object Undefined]",symToStringTag=_Symbol?_Symbol.toStringTag:void 0,_baseGetTag=baseGetTag,isObject_1=isObject,asyncTag="[object AsyncFunction]",funcTag="[object Function]",genTag="[object GeneratorFunction]",proxyTag="[object Proxy]",isFunction_1=isFunction,_createClass=function(){function e(e,o){for(var t=0;t<o.length;t++){var n=o[t];n.enumerable=n.enumerable||!1,n.configurable=!0,"value"in n&&(n.writable=!0),Object.defineProperty(e,n.key,n)}}return function(o,t,n){return t&&e(o.prototype,t),n&&e(o,n),o}}(),GooglePlaces=function(){function e(){_classCallCheck(this,e)}return _createClass(e,[{key:"currentPlace",value:function(e,o){cordova.exec(e,function(e){return o(new Error(e))},"GooglePlaces","currentPlace",[])}},{key:"autocompleteQuery",value:function(){for(var e=[],o=[],t=arguments.length,n=Array(t),r=0;r<t;r++)n[r]=arguments[r];var a=!0,c=!1,l=void 0;try{for(var i,u=n[Symbol.iterator]();!(a=(i=u.next()).done);a=!0){var s=i.value;isFunction_1(s)?o.push(s):e.push(s)}}catch(e){c=!0,l=e}finally{try{!a&&u.return&&u.return()}finally{if(c)throw l}}var g=function(){},f=function(){};o.length>0&&(g=o[0],o.length>0&&(f=function(e){return o[1](new Error(e))})),e.length>3||o.length>2?f(new Error("GooglePlaces: wrong arguments for autocompleteQuery(query, bounds, filter, success, failure)")):cordova.exec(g,f,"GooglePlaces","autocompleteQuery",e)}},{key:"pickPlace",value:function(){for(var e=[],o=[],t=arguments.length,n=Array(t),r=0;r<t;r++)n[r]=arguments[r];var a=!0,c=!1,l=void 0;try{for(var i,u=n[Symbol.iterator]();!(a=(i=u.next()).done);a=!0){var s=i.value;isFunction_1(s)?o.push(s):e.push(s)}}catch(e){c=!0,l=e}finally{try{!a&&u.return&&u.return()}finally{if(c)throw l}}var g=function(){},f=function(){};o.length>0&&(g=o[0],o.length>0&&(f=function(e){return o[1](new Error(e))})),e.length>1||o.length>2?f(new Error("GooglePlaces: wrong arguments for pickPlace(bounds, success, failure)")):cordova.exec(g,f,"GooglePlaces","pickPlace",[])}},{key:"showPlaceAutocomplete",value:function(e,o){cordova.exec(e,function(e){return o(new Error(e))},"GooglePlaces","showPlaceAutocomplete",[])}}]),e}(),AutocompleteFilterTypes={NoFilter:"no_filter",Geocode:"geocode",Address:"address",Establishment:"establishment",Region:"region",City:"city"};module.exports=new GooglePlaces,module.exports.AutocompleteFilterTypes=AutocompleteFilterTypes;


www/GooglePlaces.js
import isFunction from "lodash/isFunction";

//
// # GooglePlaces
//
// Wraps the (non-UI) methods of the Google Places SDK to Javascript.
//
class GooglePlaces {
  // ## currentPlace
  //
  // `currentPlace`(`sucess`, `failure`)
  //
  // Discover the place where the device is currently located.
  //
  // ### Important notice
  //
  // This method *requires* that the user has enabled geolocation in the app.
  // To to so, use the [dedicated cordova plugin](https://github.com/apache/cordova-plugin-geolocation) before calling this method.
  //
  // ### Parameters
  //
  // - `success` is called in case of success, it will contain "place likehood" objects
  //   with a `place` and `likehood` fields:
  //   ```javascript
  //   {
  //    place: {
  //      name: "some place name",
  //      placeID: "XXXXX"
  //    },
  //    likehood: 0.87 // <= means 87% accurate
  //   }
  //   ```
  // - `failure` is called in case of an error, with an error objects
  //
  currentPlace(success, failure) {
    cordova.exec(
      success,
      err => failure(new Error(err)),
      "GooglePlaces",
      "currentPlace",
      [],
    );
  }

  // ## autocompleteQuery
  //
  // `autocompleteQuery`(`query`, `[bounds]`, `[filter]`, `success`, `[failure]`)
  //
  // Runs a query to offer auto-completion results from a query.
  //
  // #### Note
  //
  // this method takes a *variable* number of arguments.
  //
  // ### Parameters
  //
  // - `query`: the actual query: an incomplete address.
  // - `bounds`: a region to limit the search to.
  //
  //    It should be defined as a "coordinate region" object such as:
  //
  //    ```javascript
  //    {
  //     northEast: {
  //       latitude: 1.234,
  //       longitude: 5,667
  //     },
  //     southWest: {
  //       latitude: 1.234,
  //       longitude: 5,667
  //     }
  //    }
  //    ```
  //
  // - `filter`: a filter to limit the results to a specific region.
  //
  //   Such a filter is given by a filter type taken from the `GooglePlaces.AutocompleteFilterTypes` and an (optional) country:
  //
  //   ```javascript
  //   {
  //     filter: AutocompleteFilterTypes.NoFilter,
  //     country: "FR" // <= this is optional
  //   }
  //   ```
  //
  // - `success` is called in case of success, it will contain "autocomplete prediction" objects
  //   with info fields:
  //   ```javascript
  //   {
  //     fullText: "description of the place",
  //     primaryText: "partial description of the place",
  //     secondaryText: "partial description of the place",
  //     placeID: "XXXXX",
  //     types: [ "a", "list", "of", "types", "for", "the", "result" ]
  //   }
  //   ```
  // - `failure` is called in case of an error, with an error object.
  autocompleteQuery(...args) {
    let params = [];
    let callbacks = [];

    for (let arg of args) {
      if (isFunction(arg)) {
        callbacks.push(arg);
      } else {
        params.push(arg);
      }
    }

    let success = () => {};
    let failure = () => {};
    if (callbacks.length > 0) {
      success = callbacks[0];
      if (callbacks.length > 0) {
        failure = err => callbacks[1](new Error(err));
      }
    }

    if (params.length > 3 || callbacks.length > 2) {
      const err = new Error(
        "GooglePlaces: wrong arguments for autocompleteQuery(query, bounds, filter, success, failure)",
      );
      failure(err);
      return;
    }

    cordova.exec(success, failure, "GooglePlaces", "autocompleteQuery", params);
  }

  // ## showPlaceAutocomplete
  //
  // `showPlaceAutocomplete`(`sucess`, `failure`)
  //
  // Show the native UI for Picking a nearby place
  //
  // ### Important notice
  //
  // This method *requires* that the user has enabled geolocation in the app.
  // To to so, use the [dedicated cordova plugin](https://github.com/apache/cordova-plugin-geolocation) before calling this method.
  //
  // ### Parameters
  //
  // - `bounds`: (optinal) a region to limit the search to.
  //
  //    It should be defined as a "coordinate region" object such as:
  //
  //    ```javascript
  //    {
  //     northEast: {
  //       latitude: 1.234,
  //       longitude: 5,667
  //     },
  //     southWest: {
  //       latitude: 1.234,
  //       longitude: 5,667
  //     }
  //    }
  //    ```
  //
  // - `success` is called in case of success, it will contain "place" objects
  //   with a `place` and `likehood` fields:
  //   ```javascript
  //   {
  //      name: "some place name",
  //      placeID: "XXXXX",
  //      // and lots of other fields, depending on the place info available
  //   }
  //   ```
  // - `failure` is called in case of an error, with an error objects
  //
  pickPlace(...args) {
    let params = [];
    let callbacks = [];

    for (let arg of args) {
      if (isFunction(arg)) {
        callbacks.push(arg);
      } else {
        params.push(arg);
      }
    }

    let success = () => {};
    let failure = () => {};
    if (callbacks.length > 0) {
      success = callbacks[0];
      if (callbacks.length > 0) {
        failure = err => callbacks[1](new Error(err));
      }
    }

    if (params.length > 1 || callbacks.length > 2) {
      const err = new Error(
        "GooglePlaces: wrong arguments for pickPlace(bounds, success, failure)",
      );
      failure(err);
      return;
    }

    cordova.exec(success, failure, "GooglePlaces", "pickPlace", []);
  }

  // ## showPlaceAutocomplete
  //
  // `showPlaceAutocomplete`(`sucess`, `failure`)
  //
  // Show the native UI for Place Autocomplete
  //
  // ### Parameters
  //
  // - `success` is called in case of success, it will contain "place" objects
  //   with a `place` and `likehood` fields:
  //   ```javascript
  //   {
  //      name: "some place name",
  //      placeID: "XXXXX",
  //      // and lots of other fields, depending on the place info available
  //   }
  //   ```
  // - `failure` is called in case of an error, with an error objects
  //
  showPlaceAutocomplete(success, failure) {
    cordova.exec(
      success,
      err => failure(new Error(err)),
      "GooglePlaces",
      "showPlaceAutocomplete",
      [],
    );
  }
}

// ## Filters for the `autocompleteQuery` method.
const AutocompleteFilterTypes = {
  // - `AutocompleteFilterTypes.NoFilter` is an empty filter; all results are returned.
  NoFilter: "no_filter",
  // - `AutocompleteFilterTypes.Geocode` returns only autocomplete results with a precise address. Use this type when you know the user is looking for a fully specified address.
  Geocode: "geocode",
  // - `AutocompleteFilterTypes.Address` returns only places that are businesses.
  Address: "address",
  // - `AutocompleteFilterTypes.Establishment` returns only places that are businesses.
  Establishment: "establishment",
  // - `AutocompleteFilterTypes.Region` returns only places that match one of the following types: `locality`, `sublocality`, `postal_code`, `country`, `administrative_area_level_1`, `administrative_area_level_2`
  Region: "region",
  // - `AutocompleteFilterTypes.City` returns only results matching `locality` or `administrative_area_level_3`.
  City: "city",
};

module.exports = new GooglePlaces();
module.exports.AutocompleteFilterTypes = AutocompleteFilterTypes;
